docs(Debug SPL): Updates U-Boot SPL Debug Guide (#272)
TexasInstruments/processor-sdk-doc: source/linux/Foundational_Components/U-Boot/Apps-SPL-Debug-OpenOCD.rst

------------------------
Debugging U-Boot and SPL
------------------------

Debugging SPL is important for bug fixes, new features, and better understanding of
the boot process. The AM62x primary boot flow starts on TIFS, which boots R5 and later A53. Refer to

> `3.1.1.1.4. Boot Flow <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Foundational_Components/U-Boot/UG-General-Info.html#boot-flow>`_

Reviewer: Use `:ref` tag for cross referencing sections which are from the same processor-sdk-doc project

https://www.sphinx-doc.org/en/master/usage/referencing.html#role-ref

Author: Resolved this and other instances by adding ref labels at referenced locations

> for more info about the AM62x boot flow. OpenOCD/GDB and CCS are useful tools to debug source

Author: This guide shows debugging for both OpenOCD and CCS

code modifications at very early stages of the boot process, even at initial chip reset when
nothing has printed to the console by the device. The below example is useful to understand how to
properly debug at these early stages.

.. note::

   This guide uses the AM62x, though the process is very similar to other TI K3 devices.

^^^^^^^^^^^^^
Debug Example
^^^^^^^^^^^^^

Use OpenOCD/GDB and CCS on a Linux host machine to debug SPL on the SK-AM62B-P1 board,
either in the *board_init_f()* function or upon R5 or A53 processor reset.

.. note::

   First, ensure you're able to:

>       - Clone, access, and modify the *ti-u-boot* source code: `3.1.1.1.1. Getting the U-Boot Source Code <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Foundational_Components/U-Boot/UG-General-Info.html#getting-the-u-boot-source-code>`_

Reviewer: Use `:ref` tag for cross referencing sections which are from the same processor-sdk-doc project

https://www.sphinx-doc.org/en/master/usage/referencing.html#role-ref

      - Build u-boot: `3.1.1.1.2. Build U-Boot <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Foundational_Components/U-Boot/UG-General-Info.html#build-u-boot>`_ or `1.1.4. SDK Build using Makefile <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Overview/Top_Level_Makefile.html#sdk-build-using-makefile>`_
      - Create an SD card: `1.1.2. Create SD Card <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/Overview/Processor_SDK_Linux_create_SD_card.html#create-sd-card>`_
      - Boot up and interface with the SK-AM62B-P1 board: `4.2.1. AM62x SK EVM Hardware Setup <https://software-dl.ti.com/processor-sdk-linux/esd/AM62X/latest/exports/docs/linux/How_to_Guides/Hardware_Setup_with_CCS/AM62x_EVM_Hardware_Setup.html#am62x-sk-evm-hardware-setup>`_

   before continuing with the below debugging example.

#. Add a debug dead loop:

>    **Option A**: in the *board_init_f()* function in the *am625_init.c* source code file:

Reviewer: ```suggestion
   **Option A**: in the *board_init_f()* function in the :file:`am625_init.c` source code file:
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

>       #. Open the *am625_init.c* source code at the file path:

Reviewer: ```suggestion
      #. Open the :file`am625_init.c` source code at the file path:
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

>          .. code-block:: console

Reviewer: code-block isn’t applicable here...you can just use the file role to highlight the below file-path

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

Author: Resolved this and other instances

            ti-u-boot/arch/arm/mach-k3/am62x/am625_init.c

      #. Add a debug dead loop in *board_init_f()*:

         .. code-block:: c

            void board_init_f(ulong dummy)
            {
               .
               .
               store_boot_info_from_rom();

               ctrl_mmr_unlock();

               +asm("test: nop");
               +asm("       nop");
               +asm("  b test ");
               +asm(" nop");
               +asm(" nop");
               +
               +if (IS_ENABLED(CONFIG_CPU_V7R)) {
               +    volatile int x = 1;
               +    while(x) {};
               +}
               +
               +if (IS_ENABLED(CONFIG_ARM64)) {
               +    volatile int x = 1;
               +    while(x) {};
               +}

               spl_early_init();
               .
               .

>             }

Author: As also shown in that link, the purpose is "to catch the CPU at the desired place during the bootup process", so the user can modify the PC (or variable for infinite while-loop) and step through the code for debug

         - These dead loops are in both in-line assembly and C. The first dead loop (shown with *asm*
           in-line assembly) will run on every processor. The second dead loop (shown as two if
           statement blocks) will only run on the specific processor: R5 *(CONFIG_CPU_V7R)* or
           A53 *(CONFIG_ARM64)*.

   **Option B**: at R5 or A53 processor *reset* in the *start.S* source code file:

      #. Open the *start.S* source code at the file path:

         **R5:**

>          .. code-block:: console

Reviewer: Same comment as earlier

      #. Add a debug dead loop at *reset*:

         **R5:**

         .. code-block:: c

            .
            .
            #ifdef CONFIG_ARMV7_LPAE
               .global	switch_to_hypervisor_ret
            #endif

            reset:
            +dead_loop:
            +	nop
            +	b	dead_loop
               /* Allow the board to save important registers */
               b	save_boot_params
            save_boot_params_ret:
            #ifdef CONFIG_POSITION_INDEPENDENT
            .
            .

         **A53:**

         .. code-block:: c

            .
            .
            .globl	_bss_end_ofs
            _bss_end_ofs:
               .quad	__bss_end - _start

            reset:
            +dead_loop:
            +	nop
            +	b	dead_loop
               /* Allow the board to save important registers */
               b	save_boot_params
            .globl	save_boot_params_ret
            save_boot_params_ret:
            .
            .

         - These dead loops are in assembly and are specific to the R5 or A53 processor in
           the AM62x, based on which *start.S* file you opened and modified above.

   .. note::

>       On AM335x/AM437x/AM57xx, a debug dead loop could be added in the *early_system_init* function at

Author: Updated with ifconfigs to only show this note and other instances for those devices

>       the file path *ti-u-boot/arch/arm/mach-omap2/hwinit-common.c*. The location of a debug dead loop

Reviewer: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

      on all platforms will vary based on the section of the SPL we are trying to debug.

#. Move the modified u-boot binaries to the SD card:

   After building u-boot, copy/replace the *tiboot3.bin*, *tispl.bin*, and *u-boot.img*
   u-boot binaries to the SD card's boot partition and insert the SD card in the board.

   .. note::

>       On AM335x/AM437x/AM57xx, copy *MLO* and *u-boot.img* to the SD card. On AM65x, copy

Reviewer: New line space between note & the content block

Author: Added

>       On AM335x/AM437x/AM57xx, copy *MLO* and *u-boot.img* to the SD card. On AM65x, copy

Reviewer: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

      *sysfw.itb*, *tiboot3.bin*, *tispl.bin* and *u-boot.img* to the SD card.

#. Power on the SK-AM62B-P1 board:

   With a micro-USB cable, plug in the board to the host machine via the JTAG
   XDS110 port:

   .. Image:: /images/sk-am62b-p1-top-jtag.png
      :width: 625

   With a USB-C cable, plug in the board to power using either USB-C port:

   .. Image:: /images/sk-am62b-p1-top-power.png
      :width: 625

"""""""""""""""""""
Using OpenOCD & GDB

> """""""""""""""""""

Reviewer: This isnt a suggested convention of section headers from sphinx

Refer https://lpn-doc-sphinx-primer.readthedocs.io/en/stable/concepts/heading.html

Reviewer 2: @cshilwant that is not the correct specification. We follow https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections

Of which, this PR is already aligned. Now, we could argue that this document section should start at ordinal 0 (`#` with overline, for parts).

OpenOCD and GDB are useful command-line tools used for debugging on embedded systems. Here,
OpenOCD communicates with JTAG, and GDB communicates with OpenOCD.

> #. Download, install, and setup OpenOCD and GDB-Multiarch: https://docs.u-boot.org/en/latest/board/ti/k3.html#step-1-download-and-install-openocd

Reviewer: Refer https://lpn-doc-sphinx-primer.readthedocs.io/en/stable/concepts/heading.html to see how to Link to External Web Pages

Author: My original intent was to have it shown as the full link without link text. Assuming you mean to have link text for external links, I have resolved this and other instances by adding link text:
https://lpn-doc-sphinx-primer.readthedocs.io/en/stable/concepts/hyperlink.html.

         (gdb) tui enable
         (gdb) layout src
         (gdb) <Ctrl + X>
         <2>
         (gdb) <Ctrl + X>
         <2>
         (gdb) next
         (gdb) step
         (gdb) info registers pc
         pc             0x80080028             0x80080028 <reset>
         (gdb) set $pc += 8
         (gdb) info registers pc
         pc             0x80080030             0x80080030 <reset+8>
         (gdb) b save_boot_params
         Breakpoint 1 at 0x800800e8: file /home/${USER}/ti-u-boot/arch/arm/cpu/armv8/start.S, line 399.
         (gdb) continue
         Continuing.

         Breakpoint 1, save_boot_params () at /home/${USER}/ti-u-boot/arch/arm/cpu/armv8/start.S:399
         (gdb) next
         save_boot_params_ret () at /home/${USER}/ti-u-boot/arch/arm/cpu/armv8/start.S:125
         (gdb) next
         (gdb) nexti
         (gdb) tui disable
         (gdb) q
         A debugging session is active.

            Inferior 1 [Remote target] will be detached.

         Quit anyway? (y or n) y
         Detaching from pid Remote target
         [Inferior 1 (Remote target) detached]

      - An easy way to break the loop is to change the program counter.

> Refer to the U-Boot K3 OpenOCD/GDB documentation for more info: https://docs.u-boot.org/en/latest/board/ti/k3.html#common-debugging-environment-openocd.

Author: Good idea, will consider that for future docs modification

> """""""""

Reviewer: Same comment of section headers as above

Using CCS

> """""""""

Author: Since it's using 2 different debug tools under the same debug guide, it probably wouldn't be ideal to break up the guide across multiple pages

CCS is a useful GUI IDE with a built-in debugger for embedded systems. To debug, CCS communicates with JTAG.

> #. Download, install, setup, and open CCS (Linux): https://www.ti.com/tool/download/CCSTUDIO.

Reviewer: Refer https://lpn-doc-sphinx-primer.readthedocs.io/en/stable/concepts/heading.html to see how to Link to External Web Pages

   #. After successful launch, connect to one of the ARM cores:

      .. Image:: /images/Uboot_Image7.png
         :width: 850

      .. note::

         On AM62x, the inital SPL code is executed on the R5, so the user would connect
         to the R5 core. On AM335x/AM437/AM57xx, connect to the A8/A9/A15 core instead.

   #. After connecting to the core:

      - Click on "Run" -> "Load" -> "Load Symbols...":

         .. Image:: /images/Uboot_Image8.png
            :width: 850

      - Click "Browse":

         .. Image:: /images/Uboot_Image9.png
            :width: 500

      - Click "u-boot-spl" with no file extension (you may have to change the file type to "All Files"):

         .. Image:: /images/Uboot_Image10.png
            :width: 750

   #. After loading the symbol file, CCS should find the source code file stuck at the added debug dead loop:

      .. Image:: /images/Uboot_Image11.png
         :width: 1000

   #. To break out of the loop, change the program counter in the "Register" window:

      .. Image:: /images/Uboot_Image12.png
         :width: 1000

At this point, the user has the ability to step through the SPL code and debug issues in the SPL.

.. note::

>    On K3 family of devices such as AM62x, a watchdog timer part of the DMSC is enabled by default by the ROM bootcode

Reviewer 3: Be careful copying data from upstream sources. License in upstream is GPL-2.0+ OR BSD-3-Clause -> I would rather rewrite the TI documentation from scratch. Refer to upstream documentation by using links rather than copying content.

Author: This is one of the few notes that was present in the previous version of this CCS SPL debug guide (timeframe CCS v9, roughly 2019) and decided to keep it in case it's still useful. I couldn't find this upstream info via links. If you have any links, send my way and I'll update it

Reviewer 3: https://docs.u-boot.org/en/latest/board/ti/k3.html#stepping-through-the-code https://github.com/u-boot/u-boot/blob/master/doc/board/ti/k3.rst?plain=1#L1173

either in the *board_init_f()* function or upon R5 or A53 processor reset.

.. note::

   First, ensure you're able to:

      - Clone, access, and modify the *ti-u-boot* source code: :ref:`Getting the U-Boot Source Code-label`
      - Build u-boot: :ref:`Build-U-Boot-label`
      - Create an SD card: :ref:`processor-sdk-linux-create-sd-card`
      - Boot up and interface with the SK-AM62B-P1 board: :ref:`AM62x_EVM_Hardware_Setup-label`

   before continuing with the below debugging example.

#. Add a debug dead loop:

   **Option A**: in the *board_init_f()* function in the :file:`am625_init.c` source code file:

      #. Open the :file:`am625_init.c` source code at the file path:

         :file:`ti-u-boot/arch/arm/mach-k3/am62x/am625_init.c`

      #. Add a debug dead loop in *board_init_f()*:

         .. code-block:: c

            void board_init_f(ulong dummy)
            {
               .
               .
               store_boot_info_from_rom();

               ctrl_mmr_unlock();

               +asm("test: nop");
               +asm("       nop");
               +asm("  b test ");
               +asm(" nop");

>                +asm(" nop");

Reviewer 4: ah sorry I think I may have added to the confusion here. pre C runtime all we have is assembly so that branch to self is useful. But once we have that we can use plain C

What I was trying to describe was when you're using any debugger scripting and want to catch the core it's helpful to do something like this:

```c
if (IS_ENABLED(CONFIG_CPU_V7R)) {
        volatile int x = 1;
        while(x)
                asm("nop");
}
```
so the `$pc` register changes inside that loop and our scripting can properly halt the core and adjust without us having to inject a halt command manually. But that's probably an advanced topic. If people are learning the debugging basics they probably aren't writing debug scripting :)

Author: Thanks for clarification Bryan. I added _asm("nop");_ inside while loop for more intuitive debug as you suggested

   After building u-boot, on AM62x, copy/replace the :file:`tiboot3.bin`, :file:`tispl.bin`, and :file:`u-boot.img`
   u-boot binaries to the SD card's boot partition and insert the SD card in the board.

   .. ifconfig:: CONFIG_part_family in ('AM335X_family', 'AM437X_family', 'AM57X_family')

      .. note::

         On AM335x/AM437x/AM57xx, copy :file:`MLO` and :file:`u-boot.img` to the SD card.

   .. ifconfig:: CONFIG_part_family in ('AM65X_family')

      .. note::

         On AM65x, copy
         :file:`sysfw.itb`, :file:`tiboot3.bin`, :file:`tispl.bin`, and :file:`u-boot.img` to the SD card.

#. Power on the SK-AM62B-P1 board:

   With a micro-USB cable, plug in the board to the host machine via the JTAG
   XDS110 port:

   .. Image:: /images/sk-am62b-p1-top-jtag.png
      :width: 625

   With a USB-C cable, plug in the board to power using either USB-C port:

   .. Image:: /images/sk-am62b-p1-top-power.png
      :width: 625

"""""""""""""""""""
Using OpenOCD & GDB
"""""""""""""""""""

OpenOCD and GDB are useful command-line tools used for debugging on embedded systems. Here,
OpenOCD communicates with JTAG, and GDB communicates with OpenOCD.

> #. `Download, install, and setup OpenOCD and GDB-Multiarch <https://docs.u-boot.org/en/latest/board/ti/k3.html#step-1-download-and-install-openocd>`_.

Reviewer 2: ```suggestion
#. `Download, install, and setup OpenOCD and GDB-Multiarch <https://docs.u-boot.org/en/latest/board/ti/k3.html#step-1-download-and-install-openocd>`__.
```

One underscore makes it a named link. Two makes it anonymous. Unless you plan on referencing the link again by name please make it anonymous.

Author: Resolved this and other instances by adding second underscore

#. Connect to the SK-AM62B-P1 board using OpenOCD (board connected via JTAG XDS110 and powered on):

   .. code-block:: console

      $ openocd -f ~/openocd/tcl/board/ti_am625evm.cfg
      Open On-Chip Debugger 0.12.0+dev-01925-g169d463a3 (2025-04-04-11:08)
      Licensed under GNU GPL v2
      .
      .
      Info : Listening on port 6666 for tcl connections
      Info : Listening on port 4444 for telnet connections
      Info : XDS110: connected
      .
      .
      Info : XDS110: connected to target via JTAG
      .
      .
      Info : JTAG tap: am625.cpu tap/device found: 0x0bb7e02f (mfg: 0x017 (Texas Instruments), part: 0xbb7e, ver: 0x0)
      Info : [am625.axi_ap] Examination succeed

>       Info : [am625.cpu.sysctrl] starting gdb server on 3333

Reviewer 4: So another thing we may want to consider mentioning is the part's security variant. Most newer SoCs will block any access to the sysctrl core and SE parts will absolutely deny any JTAG access unless we add the debug entry into the tiboot3 cert or modify some pins. 

We don't have to be that descriptive I don't think, just something along the of 'the security variant of the part will determine which core you have access to via the JTAG interface' kind of notice

Reviewer 3: It is good to be descriptive, but I wonder if that could be added to upstream uboot documentation.

------------------------
Debugging U-Boot and SPL
------------------------

Debugging SPL is important for bug fixes, new features, and better understanding of
the boot process. The AM62x primary boot flow starts on TIFS, which boots R5 and later A53. Refer to
:ref:`Boot-Flow-label` for more info. OpenOCD/GDB and CCS are useful tools to debug source
code modifications at very early stages of the boot process, even at initial chip reset when
nothing has printed to the console by the device. The below example is useful to understand how to
properly debug at these early stages.

.. note::

   This guide uses the AM62x, though the process is very similar to other TI K3 devices.

^^^^^^^^^^^^^
Debug Example
^^^^^^^^^^^^^

> Use OpenOCD/GDB and CCS on a Linux host machine to debug SPL on the SK-AM62B-P1 board,
> either in the *board_init_f()* function or upon R5 or A53 processor reset.
> 

Reviewer 2: ```suggestion
The following uses OpenOCD/GDB and CCS on a Linux host machine to debug SPL on the SK-AM62B-P1 board,
hooking in to either the *board_init_f()* function or upon R5 or A53 processor reset.

```

This appears to be an example summary, where the actual usage of OpenOCD and GDB are split into their own subsection. That's good, but please tag the setup section accordingly.

Author: Resolved by adding "Setup" section at this spot in the rst

               +if (IS_ENABLED(CONFIG_ARM64)) {
               +    volatile int x = 1;
               +    while(x) {
               +        asm("nop");
               +    }
               +}

               spl_early_init();
               .
               .
            }

         - These dead loops are in both in-line assembly and C. The first dead loop (shown with *asm*
           in-line assembly) will run on every processor. The second dead loop (shown as two if
           statement blocks) will only run on the specific processor: R5 *(CONFIG_CPU_V7R)* or
           A53 *(CONFIG_ARM64)*.

   **Option B**: at R5 or A53 processor *reset* in the :file:`start.S` source code file:

      #. Open the :file:`start.S` source code at the file path:

         **R5:**

         :file:`ti-u-boot/arch/arm/cpu/armv7/start.S`

         **A53:**

         :file:`ti-u-boot/arch/arm/cpu/armv8/start.S`

      #. Add a debug dead loop at *reset*:

         **R5:**

         .. code-block:: c

            .
            .
            #ifdef CONFIG_ARMV7_LPAE
               .global	switch_to_hypervisor_ret
            #endif

            reset:
            +dead_loop:
            +	nop

>             +	b	dead_loop

Reviewer 2: I missed this earlier, but is this supposed to be a diff? We support the diff syntax highlighting scheme if you migrate to a proper diff format.

Author: It's supposed to show the dead loop added to the src code. Originally I wanted the code block to show the C/Assembly syntax highlighting, but using diff is a more optimal solution. It's been updated to use diff instead, thanks
